// _vcs-videoplayer/videoplayer.js
'use strict'
/******************************************************************************
  STACK BASE
******************************************************************************/
const { h, hsvg } = require('_h')
const s           = require('_s')
/******************************************************************************
  CUSTOM
******************************************************************************/
var script = document.querySelector('#wistiascript')
if (!script) {
  script = document.createElement('script')
  script.src = '//fast.wistia.com/assets/external/E-v1.js'
  script.async = true
  script.id = 'wistiascript'
  document.head.appendChild(script)
  var _wq = window._wq = window._wq || []
}
/******************************************************************************
  MAIN

  Currently supports only "wistia videos" given a wistia id

  @TODO: check https://github.com/excid3/wistia-player-speed to extend

  @TODO: check wistia video api: http://wistia.com/doc/player-api#methods

  // <iframe src="http://fast.video.net/embed/iframe/rmhd84p5to?videoFoam=true" allowtransparency="true" frameborder="0" scrolling="no" allowfullscreen mozallowfullscreen webkitallowfullscreen oallowfullscreen msallowfullscreen width="640" height="360" videoFoam=true"></iframe>

  // <div class="wistia_embed wistia_async_rmhd84p5to" style="width:640px;height:360px;" videoFoam=true"></div>
******************************************************************************/
module.exports = videoplayer

  var x = s(`
    video {
      width: 100%;
      height: 100%;
      border: 0;
    }
  `)

function videoplayer (id, opts, callback) {
  var protocol = opts.protocol || ''
  return h('iframe', {
    src: protocol+'//fast.wistia.net/embed/iframe/' + id,
    className: 'wistia_embed' + ' ' + x('video'),
    hook: function (node) {
      var video = {}
      video[id] = callback
      _wq.push(video)
    }
  })
}
