// _vcs-promocode/promocode.js
'use strict'
/******************************************************************************
  STACK BASE
******************************************************************************/
var { h, hsvg } = require('_h')
var s           = require('_s')
/******************************************************************************
  STACK EXTEND
******************************************************************************/
var path        = require('path')
var urify       = require('urify')
var minixhr     = require('minixhr')
var defined     = require('defined')
var after       = require('after')
/******************************************************************************
  CUSTOM
******************************************************************************/
var auth = require('_auth')
var githubapi = require('_githubapi')
/******************************************************************************
  ASSETS
******************************************************************************/
var start = urify(path.join(__dirname, 'start.png'))
var header = urify(path.join(__dirname, 'header.png'))
var closebutton = urify(path.join(__dirname, 'closebutton.png'))
/******************************************************************************
  MAIN
******************************************************************************/
module.exports = PromocodeComponent

function PromocodeComponent (db, engine) { // SIGNUP COMPONENT - OVERLAY


  /*****************************/
  // THEME + LOCAL CACHE + DEFAULT VALUES
  var esovablue = '#526E96'
  var esovalightblue = '#edf2fa'
  var esovalightgreen = '#35E38A'
  var esovared = 'FA9B68'
  var esovafont1 = 'Avenir Roman'
  var FREE_LESSONS  = 4
  var STATUS = {
    message: 'clear',
    promocode: ''
  }
  var api = githubapi(db)

  /*****************************/
  // INITIALIZE STYLING
  var x = s(`
    promocode {
      display: flex;
      flex-direction: column;
      height: 100%;
      width: 100%;
      justify-content: flex-start;
      align-items: center;
      z-index: 10;
      background-color: hsla(0, 0%, 0%, 0.19);
    }
    welcome {
      position: relative;
      margin-top: 253px;
      display: flex;
      box-sizing: content-box;
      flex-direction: column;
      justify-content: flex-start;
      align-items: center;
      min-width: 200px;
      width: 80%;
      max-width: 555px;
      min-height: 365px;
      height: 27%;
      padding: 0 10px 80px 10px;
      background-color: hsla(0, 0%, 100%, 1);
      color: black;
      font-family: ${esovafont1};
      border-radius: 15px;
      font-size: 3vmin;
    }
    message {
      color: ${esovalightgreen};
      font-size: 18px;
      padding-bottom: 10px;
      line-height: 30px;
      text-align: center;
      font-weight: normal;
      font-family: ${esovafont1};
    }
    haveanaccount {
      color: ${esovablue};
      font-size: 14px;
      text-decoration: underline;
      padding-top: 10px;
      line-height: 30px;
      text-align: center;
      font-weight: normal;
      font-family: ${esovafont1};
      cursor: pointer;
    }
    unmetrequirements {
      color: ${esovared};
      font-size: 0.8em;
      text-align: center;
      font-weight: normal;
      font-family: ${esovafont1};
      margin-top: 0px;
      padding: 80px 115px 40px 115px;
    }
    codeinput {
      padding-top: 55px;
    }
    errormessage {
      color: ${esovared};
      font-size: 18px;
      line-height: 30px;
      text-align: center;
      font-weight: normal;
      font-family: ${esovafont1};
    }
    promoinput {
      background-color: ${esovalightblue};
      font-size: 18px;
      padding: 10px;
      border: 1px solid ${esovablue};
      border-top-left-radius: 15px;
      border-bottom-left-radius: 15px;
      color: ${esovablue};
      border-right: 0;
      font-family: ${esovafont1};
      text-align: center;
    }
    signup {
      background-color: ${esovablue};
      border: 1px solid ${esovablue};
      border-top-right-radius: 15px;
      border-bottom-right-radius: 15px;
      padding: 10px 15px;
      font-family: ${esovafont1};
      color: white;
      font-size: 18px;
      cursor: pointer;
    }
    notready {
      display: flex;
      align-items: center;
      justify-content: center;
    }
    startnow {
      background-color: ${esovablue};
      border: 1px solid ${esovablue};
      border-radius: 15px;
      padding: 3px 15px;
      font-family: ${esovafont1};
      color: white;
      font-size: 18px;
      cursor: pointer;
    }
    closebutton {
      align-self: flex-end;
      padding-top: 10px;
      cursor: pointer;
    }
    welcome-text {
      color: ${esovablue};
      width: 100px;
      text-align: center;
      position: absolute;
      left: 235px;
      font-size: 25px;
      top: 50px;
    }
    speedlimit {
      color: ${esovared};
      font-size: 0.8em;
      text-align: center;
      font-weight: normal;
      font-family: ${esovafont1};
      margin-top: 0px;
      padding: 80px 115px 40px 115px;
    }
  `)


  /*****************************/
  // CHILD COMPONENTS & USED MODULES
  // ...
  function returnNotReady () {
    return h('div', [
      h('h2', { className: x('message') }, LOCALE['text/overlay/promocode:notready']
      ),
      h('div',  { className: x('notready') }, [
        h('button', {
          className: x('startnow'),
          onclick: go2currentlesson
        },
          LOCALE['text/overlay/promocode/button:startnow']
        )
      ]),
      h('h2', {
        className: x('haveanaccount'),
        onclick: onsignin
      }, LOCALE['text/overlay/promocode:haveanaccount']
      )
    ])
  }

  function returnWrongPromocode () {
    return h('div', [
      h('h2', { className: x('errormessage') },
        LOCALE['text/overlay/promocode:wrong']
      )
    ])
  }

  function returnUnmetRequirements () {
    return h('div', [
      h('h2', { className: x('unmetrequirements') },
        LOCALE['text/overlay/promocode:unmetrequirements']
      )
    ])
  }

  function returnSpeedLimit () {
    return h('div', [
      h('h2', { className: x('speedlimit') },
        LOCALE['text/overlay/promocode:speedlimit']
      )
    ])
  }

  /*****************************/
  // RENDER COMPONENT
  function renderPromocode () {
    var vtree = STATUS.message === 'clear' ?
      '' // renders nothing to clear the overlay
    : h('div', {
      className: x('promocode'),
      onclick: function (event) {
        if (this === event.target) onclose(event)
      }
    }, [
      h('div', { className: x('welcome') }, [
        h('img', {
          src: closebutton,
          className: x('closebutton'),
          onclick: onclose
        }),
        h('div', { className: x('welcome-text') }, LOCALE['text/overlay/promocode/welcome']),
        h('img', { src: header, className: x('header-picture')}),
        STATUS.message === 'not ready' ?
        returnNotReady()
        : STATUS.message === 'unmet requirements' ?
        returnUnmetRequirements()
        : STATUS.message === 'speed limit' ?
        returnSpeedLimit()
        : h('div', { className: x('codeinput') }, [
          h('input', {
            className: x('promoinput'),
            name: 'promocode',
            onchange: function (event) {
              STATUS.promocode = event.target.value
            },
            onkeyup: function (event) {
              var key = event.which
              if (key === 13) { // enter
                db.put('esova-promocode', STATUS.promocode, function (error) {
                  // auth === ACTION_signIn(...)
                  // param is/was set via db.put ...
                  auth(db, AUTHED)
                })
              }
            },
            placeholder: LOCALE['text/overlay/promocode/placeholder']
          }),
          h('button', {
            className: x('signup'),
            onclick: function () {
              // @IDEA
              // The "SIGNAL" is like a "BEHAVIOR" instead of "EVENT"
              // => when it changes, listeners are informed
              // => the signal (=value of the "event") stays the same until
              //    its changed to something else
              db.put('esova-promocode', STATUS.promocode, function (error) {
                // auth === ACTION_signIn(...)
                // param is/was set via db.put ...
                auth(db, AUTHED)
              })
            },
          }, LOCALE['text/overlay/promocode/button:signup'])
        ]),
        STATUS.message === 'no valid promocode' ?
        returnWrongPromocode()
        : ''
      ])
    ])
    // PUSH UPDATE
    engine.push({
      type: 'vtree',
      key: 'promocode',
      value: vtree
    })
  }


  /*****************************/
  // INCOMING COMMUNICATION CHANNELS
  db.on('put', function (key, value) {
    try { value = JSON.parse(value) } catch (e) { value = defined(value, {}) }

    if (key === 'esova-credentials' || key === 'OVERLAY') {
      // SELECT
      var status = value.status
      // RE-INTERPRET (based on context knowledge)
      if (status === 'logout') status = 'clear'
      if (status !== 'clear')
        db.get('esova-currentLesson', function (error, lastlesson) {
          if (+lastlesson <= FREE_LESSONS && status !== 'unmet requirements') {
            EVALUATE('not ready') // AUTH FAIL + currentLesson <= FREE_LESSONS
          }
          else if (status==='no valid promocode' && STATUS.message==='clear')
            EVALUATE('ok') // login first time fail => dont say wrong promo yet
          else EVALUATE(status)
        })
      else EVALUATE(status)
      // EVALUATE
      function EVALUATE (status) {
        var setERR = { }
        var setRENDER = {
          'ok': true,
          'no valid promocode': true,
          'not ready': true,
          'speed limit': true,
          'unmet requirements': true,
          'clear': true
        }
        if (setRENDER[status]) {
          STATUS.message = status
          renderPromocode()
        } else { // setERR or others
          console.error('STATUS '+value+' is unknown')
          debugger; db.put('ERROR', 'OVERLAY: '+status, ER)
        }
      }
    }
    /********************************/
  })


  /*****************************/
  // INITIALIZE
  console.log('---------------------------------')
  console.log('INITIALIZE: promocode() component')
  console.log('---------------------------------')
  var LOCALE = {
    "text/overlay/promocode/button:startnow": "",
    "text/overlay/promocode/button:signup": "",
    "text/overlay/promocode/placeholder": "",
    "text/overlay/promocode:notready": "",
    "text/overlay/promocode:haveanaccount": "",
    "text/overlay/promocode:wrong": "",
    "text/overlay/promocode:unmetrequirements": "",
    "text/overlay/promocode:speedlimit": "",
    "text/overlay/promocode/welcome": ""
  }

  var FIELDS = Object.keys(LOCALE).length
  var init = after(FIELDS, function (err) {
    if (err) db.put('ERROR', err)
    else {
      renderPromocode()
      init = renderPromocode
    }
  })
  db.on('put', function (key, value) {
    try { value = JSON.parse(value) } catch (e) { value = defined(value, {}) }

    if (LOCALE[key] !== undefined) {
      LOCALE[key] = value
      init()
    }
  })
  // components at least do not return vtree ...but maybe some kind of API
  // ... to set "default wireup" in context of users instantiating "usage code"
  return undefined // components don't return anything! @TODO: remove this line


  /*****************************/
  // HELPER
  function ER (error) {
    // @TODO
    // default error logging failed
    // => try to learn about error via ajax or some other technique
    // => if it fails inform user instead and hope he reports it in chat
    if (error) {
      console.error('couldnt store ERROR event')
      console.error(error)
    }
  }
  // ACTIONS
  function go2currentlesson (event) {
    event.preventDefault()
    onclose(event)
    db.get('esova-currentLesson', function (error, value) {
      if (error) value = '0'
      db.put('esova-lesson', value, function noop () { })
      db.put('MODE', 'skillbox')
    })
  }
  function onsignin (event) {
    event.preventDefault()
    auth(db,{force:true},AUTHED)
  }
  function onclose (event) {
    // basically are:
    // PRIVATE
    // - functions if they manipulate only the component itself
    // PUBLIC
    // - db.put signals if they want to inform the world too
    // - db put signals if they manipulate themselves
    //   (they just need to listen to themselves),
    //   but want to inform the world too
    event.preventDefault()
    db.put('OVERLAY', JSON.stringify({ status: 'clear' }))
  }
  function AUTHED (error, help) {
    if (error) console.error('AUTH FAILED')
    else api('getlessonNumber', function (error, number) { })
  }
}
