// _vcs-skilltree/__skill.js
const ehs         = require('eventhandler-stream')
const { h, hsvg } = require('_h')
const s           = require('_s')

const vcs         = require('_vcs')

module.exports    = skill

function skill (db, engine) {
  const DATA = {
    src        : 'default.jpg',
    title      : 'Random Skill'
  }

  db.createReadStream({
    gte: '',
    lte: ''
  })

  db.trackValueStream()
  db.trackRangeStream()

  db.createTrackStream()

  const state$ = db.trackable()

  state$.track('active')

  const click$ = ehs(function (event) {
    var isActive = DATA['active'] === DATA['title']
    if (isActive) {
      db.put(DATA['active'], null)
    } else {
      db.put(DATA['active'], DATA['title'])
    }
    return { type: 'put', key: '', value: DATA['title'] }
  })

  click$.pipe()

  const skill$ = vcs(function template (chunk, encoding, next) {
    if (isArray(DATA[chunk.key])) { DATA[chunk.key].push(chunk.value) }
    else                          { DATA[chunk.key] = chunk.value     }
    this.push(template(DATA['src'], DATA['title']))
  })

  function template (src, title) {
    return h('div', [
      h('img', { src: DATA['src'] }),
      h('button', { onclick: click$ }, DATA['title'])
    ])
  }
  
  return skill$
}
