// _vcs-skilltree/skillbox.js
'use strict'
/******************************************************************************
  STACK BASE
******************************************************************************/
var { h, hsvg } = require('_h')
var s           = require('_s')
/******************************************************************************
  STACK EXTEND
******************************************************************************/
var path        = require('path')
var urify       = require('urify')
var after       = require('after')
var defined     = require('defined')
var moment      = require('moment')
/******************************************************************************
  CUSTOM
******************************************************************************/
var githubapi = require('_githubapi')
var auth      = require('_auth')
// var gitterchat        = require('_vcs/gitterchat')
/******************************************************************************
  ASSETS
******************************************************************************/
var forward           = urify(path.join(__dirname, 'forward.png'))
var forward_pressed   = urify(path.join(__dirname, 'forward-pressed.png'))
var forward_arrow     = urify(path.join(__dirname, 'forward-arrow.png'))
var forward_pressed   = urify(path.join(__dirname, 'forward-pressed.png'))
var back              = urify(path.join(__dirname, 'back.png'))
var back_pressed      = urify(path.join(__dirname, 'back-pressed.png'))
var back_arrow        = urify(path.join(__dirname, 'back-arrow.png'))
var back_pressed      = urify(path.join(__dirname, 'back-pressed.png'))
var playicon          = urify(path.join(__dirname, 'play-icon.png'))
/******************************************************************************
  MAIN
******************************************************************************/
module.exports = skillbox

function skillbox (db, engine) {


  /*****************************/
  // THEME + LOCAL CACHE + DEFAULT VALUES
  var VTREE = {
    lessonHeaderContent: '',
    lessonDescription: ''
  }
  var STATUS = {
    // LOCAL UI STATUS
    lessonstart: undefined, // lessonstart button hover state
    navleft: undefined,
    navright: undefined,
    // LOCAL DATA STATUS
    chaturl: 'https://gitter.im/ESova/program/~embed',
    id: undefined,
    currentlesson: '0', // esova-lesson
    lastlesson: '0', // esova-currentLesson
    maxlesson: Infinity
  }
  // @TODO: switchboarded db should recognize all automatically
  var LOCALE = { // local CACHE - initial defaults are defined in template
    "text/lesson": "Lekcija",
    "text/start": "Počnite sa lekcijom",
    "text/chat:anon": "Čet pomoć"
  }
  var FREE_LESSONS = 4
  var lightbluegrey = '#D6DBE1'
  var esovadarkgreen = '#32A971'
  var esovagreen = '#35E38A'
  var esovadarkblue = 'hsl(216, 46%, 16%)'
  var esovablue = '#526E96'


  /*****************************/
  // INITIALIZE STYLING
  var x = s(`
    skillbox {
      display: flex;
      flex-direction: column;
      min-height: 450px;
      min-width: 600px;
      width: 75%;
      max-width: 1000px;
    }
    skillbox-nav {
      display: flex;
      justify-content: center;
      margin-bottom: 20px;
      padding-top: 30px;
      padding-bottom: 30px;
    }
    skillbox-description {
      display: flex;
      flex-grow: 1;
      background-color: ${'#fff'};
      border-top-right-radius: 20px;
      border-bottom-right-radius: 20px;
      border-top-left-radius: 20px;
      border-bottom-left-radius: 20px;
      min-width: 650px;
    }
    navright {
      box-sizing: border-box;
      display: flex;
      align-items: center;
      justify-content: center;
      width: 60px;
      height: 60px;
      background-position: center;
      background-image: url(${forward});
      background-size: 100%;
      background-repeat: no-repeat;
      cursor: pointer;
    }
    navleft-placeholder {
      box-sizing: border-box;
      display: flex;
      align-items: center;
      justify-content: center;
      width: 60px;
      height: 60px;
      margin-right: 100px;
    }
    navright-placeholder {
      box-sizing: border-box;
      display: flex;
      align-items: center;
      justify-content: center;
      width: 60px;
      height: 60px;
    }
    navright--hover {
    }
    navleft--hover {
    }
    navleft {
      box-sizing: border-box;
      display: flex;
      align-items: center;
      justify-content: center;
      width: 60px;
      height: 60px;
      margin-right: 100px;
      background-position: center;
      background-image: url(${back});
      background-size: 100%;
      background-repeat: no-repeat;
      cursor: pointer;
    }
    chat {
      display: flex;
      position: relative;
      overflow: hidden;
      flex-grow: 1;
      min-width: 300px;
      margin-top: 20px;
      margin-bottom: 20px;
    }
    chatframe {
      position: absolute;
      width: 125%;
      height: 125%;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%) scale(0.8);
    }

    lesson {
      font-family: ${'Avenir Roman'};
      box-sizing: border-box;
      display: flex;
      flex-direction: column;
      align-content: stretch;
      align-items: stretch;
      background-color: ${lightbluegrey};
      min-width: 300px;
      flex-grow: 4;
      padding: 50px 10%;
      border-top-left-radius: 20px;
      border-bottom-left-radius: 20px;
    }
    lesson-header {
      align-items: flex-start;
      display: flex;
      flex-direction: row;
      flex-wrap: wrap;
    }
    lesson-id {
      display: flex;
      flex-direction: row;
      line-height: 0;
      color: ${esovagreen};
    }
    lesson-number {
      margin-left: 10px;
      color: ${esovablue};
    }
    lesson-title {
      margin-right: auto;
      padding-right: 10px;
      display: flex;
      flex-direction: column;
      padding-bottom: 5px;
    }
    lessonstart {
      position: relative;
      top: 7px;
      border-radius: 15px;
      padding: 3px;
      margin-bottom: 20px;
      padding-left: 15px;
      padding-right: 15px;
      align-self: flex-start;
      white-space: nowrap;
      justify-content: flex-end;
      text-decoration: none;
      background-color: ${esovagreen};
      border: 1px solid ${esovadarkgreen};
      color: white;
    }
    lessonstart--blocked {
      background-color: ${'#ccc'};
      border: 1px solid ${'#999'};
    }
    lessonstart--hover {
      box-shadow: inset 0px 47px 52px -60px rgba(0,0,0,0.75);
    }
    lesson-seperator {
      border: 0.5pt solid white;
      width: 100%;
    }

    chat-overlay {
      position: absolute;
      bottom: 0;
      width: 100%;
      background-color: #f9f9f9;
      height: 60px;
      z-index: 1;
    }
    chatsignin {
      position: absolute;
      top: 12px;
      left: 22px;
      font-weight: bold;
      text-transform: uppercase;
      width: 147px;
      font-size: 9px;
      letter-spacing: 2px;
      border: 0px;
      border-radius: 5px;
      background-color: ${esovagreen};
      height: 35px;
      color: white;
      cursor: pointer;
    }
    chatsignin--hover {
      background-color: ${'#1FE07C'};
    }
  `)
  /*****************************/
  // CHILD COMPONENTS & USED MODULES
  var api = githubapi(db)
  // @TODO: tell them on what db.on('put') to listen...right?
  // => it seems to be currently hard coded
  lessonDescription(db, { push: renderSkillbox })
  lessonHeaderContent(db, { push: renderSkillbox })


  /*****************************/
  // RENDER COMPONENT
  function renderSkillbox (/*error,*/ update) {
    // @TODO: ALWAYS VALIDATE STATUS before RENDERING

    var childTree = (update||{}).value
    var childName = (update||{}).key

    // MAYBE UPDATE CHILDREN
    if (childName === 'lessonDescription') {
      VTREE.lessonDescription = childTree
    }
    else if (childName === 'lessonHeaderContent') {
      VTREE.lessonHeaderContent = childTree
    }
    // RENDER
    STATUS.currentlesson = +STATUS.currentlesson
    STATUS.lastlesson = +STATUS.lastlesson
    var createdAt = STATUS.createdAt
    if (createdAt) {
      var oldDate=moment(createdAt).valueOf()
      var newDate=moment.now()
      var duration=moment.duration(newDate-oldDate).asDays()
      var maxlessons = (Math.floor((duration/7))+1)*5
      if (maxlessons < STATUS.currentlesson) {
        var limitReached = true
      }
    }
    var isAllowed = STATUS.currentlesson <= FREE_LESSONS && STATUS.currentlesson <= STATUS.lastlesson ||
      STATUS.id && STATUS.currentlesson <= STATUS.lastlesson
    // var isAllowed = true
    var hasNoNavleft = STATUS.currentlesson <= 0
    var hasNoNavright = STATUS.currentlesson >= STATUS.maxlesson

    var vtree = h('div', { className: x('skillbox') }, [
      h('div', { className: x('skillbox-nav') }, [
        hasNoNavleft ? h('span',{className:x('navleft-placeholder')}) : h('a', {
          className: x('navleft', STATUS.navleft),
          onmouseover: onmouseover.bind('navleft'),
          onmouseout: onmouseout.bind('navleft'),
          // href: '#lessonY', // @TODO: fix url
          onclick: function (event) {
            event.preventDefault()
            var nextNumber = (+STATUS.currentlesson-1)
            nextNumber = nextNumber <= 0 ? 0 : nextNumber
            db.put('esova-lesson',nextNumber+'',function(error){
              if(error) db.put('ERROR', 'skillbox navleft store curr.Less', ER)
            })
          }
        }, [
          h('img', { className: x('navarrow'), src: back_arrow })
        ]),
        hasNoNavright ? h('span',{className:x('navright-placeholder')}) : h('a', {
          className: x('navright', STATUS.navright),
          onmouseover: onmouseover.bind('navright'),
          onmouseout: onmouseout.bind('navright'),
          // href: '#lessonX', // @TODO: fix url
          onclick: function (event) {
            event.preventDefault()
            // USER ACTIONS are translated into specific "db.put's"
            // Listeners to those will then act upon user intent
            db.put('esova-lesson',(+STATUS.currentlesson+1)+'',function(error){
              if (error) db.put('ERROR','skillbox navright store cur.Less', ER)
            })
          }
        }, [
          h('img', { className: x('navarrow'), src: forward_arrow })
        ]),
      ]),
      h('div', { className: x('skillbox-description') }, [
        h('div', { className: x('lesson') }, [
          h('div', { className: x('lesson-header') }, [
            h('div', { className: x('lesson-title') }, [
              h('div', { className: x('lesson-id') }, [
                h('h1', LOCALE["text/lesson"]),
                h('h1', {className: x('lesson-number')}, STATUS.currentlesson)
              ])
            ]),
            h('a', {
              className: x('lessonstart', isAllowed?'':'lessonstart--blocked', STATUS.lessonstart),
              onmouseover: onmouseover.bind('lessonstart'),
              onmouseout: onmouseout.bind('lessonstart'),
              onclick: isAllowed ? function startlesson (event) {
                event.preventDefault()
                if (limitReached) {
                  db.put('OVERLAY', JSON.stringify({
                    status: 'speed limit'
                  }), ER)
                } else {
                  db.put('MODE', 'lesson', ER)
                }
              } : function notify (event) {
                event.preventDefault()
                db.put('OVERLAY', JSON.stringify({ status: 'unmet requirements' }), ER)
              },
              href: '/#lesson'}, [
              h('img', { src: playicon, style: { marginRight: '10px' } }),
              LOCALE["text/start"]
            ])
          ]),
                        VTREE.lessonHeaderContent,
          h('hr', { className: x('lesson-seperator')}),
          VTREE.lessonDescription
        ]),
        h('div', { className: x('chat') }, (function (thechat) {
          if (!(STATUS.id)) thechat.push(h('div', {
            className: x('chat-overlay')
          }, [
            h('button', {
              onmouseover: onmouseover.bind('chatsignin'),
              onmouseout: onmouseout.bind('chatsignin'),
              onclick: function (event) {
                auth(db, function noop (error, help) { })
              },
              className: x('chatsignin', STATUS.chatsignin)
            }, LOCALE["text/chat:anon"])
          ]))
          thechat.push(h('iframe', {
            className: x('chatframe'),
            frameBorder: '0',
            src: STATUS.chaturl }))
          return thechat
        })([]))
      ])
    ])
    // PUSH UPDATE
    engine.push({
      type: 'vtree',
      key: 'skillbox',
      value: vtree
    })
  }


  /*****************************/
  // INCOMING COMMUNICATION CHANNELS
  db.on('put', function (key, value) {
    try { value = JSON.parse(value) } catch (e) { value = defined(value, {}) }

    if (key ==='esova-credentials') {
      var ID = value.id
      STATUS.id = ID
      STATUS.createdAt = value.createdAt
      renderSkillbox()
    }
    else if (key === 'esova-currentLesson') {
      var lastlesson = value
      if (STATUS.lastlesson !== lastlesson) {
        STATUS.lastlesson = lastlesson
        renderSkillbox()
      }
    }
    else if (key === 'esova-lesson') {
      var lesson = value
      if (STATUS.currentlesson !== lesson) {
        // might change the dislayed lesson number
        STATUS.currentlesson = lesson
        updateLessonContent()
      }
    } else if (key === 'esova-maxlessons') {
      var max = value
      if (STATUS.maxlesson !== max) {
        STATUS.maxlesson = max
        if (STATUS.currentlesson > max) db.put('esova-lesson', max, ER)
        renderSkillbox()
      }
    }
  })

  /*****************************/
  // INITIALIZE
  console.log('--------------------------------')
  console.log('INITIALIZE: skillbox() component')
  console.log('--------------------------------')

    // @TODO: set by parent through subEngine // <= wat?!?
    db.get('esova-credentials', function (error, value) {
      try { value = JSON.parse(value) } catch (e) { value = defined(value, {}) }
      var credentials = value
      var ID = credentials.id
      // @TODO: use '#lessonXXX' in URL instead of 'esova-lesson'
      db.get('esova-lesson', function (no_currentlesson, currentlesson) {
        db.get('esova-currentLesson', function (no_lastlesson, lastlesson) {
          if (no_currentlesson) {
            if (no_lastlesson) {
              currentlesson = lastlesson = '0'
              db.put('esova-lesson', currentlesson)
              db.put('esova-currentLesson', lastlesson)
            } else {
              currentlesson = '0'
              db.put('esova-lesson', currentlesson)
            }
          } else {
            if (no_lastlesson) {
              lastlesson = '0'
              db.put('esova-currentLesson', lastlesson)
            }
          }
          db.get('esova-maxlessons', function (error, val) {
            var max = val <= 0 ? Infinity : val
            updateLessonContent()
            // if nothing changed, no need to re-render
            if (
              STATUS.id !== ID ||
              STATUS.lastlesson !== lastlesson ||
              STATUS.currentlesson !== currentlesson ||
              STATUS.maxlesson !== max
            ) {
              STATUS.id = ID
              STATUS.createdAt = credentials.createdAt
              STATUS.currentlesson = currentlesson
              STATUS.lastlesson = lastlesson
              STATUS.maxlesson = max
              renderSkillbox()
            }
          })
        })
      })
    })

  // @TODO: switchboarded db should recognize all automatically
  var FIELDS = Object.keys(LOCALE).length
  var init = after(FIELDS, function (err) {
    if (err) db.put('ERROR', err)
    else {
      renderSkillbox()
      init = renderSkillbox
    }
  })
  db.on('put', function (key, value) {
    try { value = JSON.parse(value) } catch (e) { value = defined(value, {}) }

    if (LOCALE[key] !== undefined) {
      LOCALE[key] = value
      init()
    }
  })


  /*****************************/
  // HELPER
  function updateLessonContent () {
    api('getlessondescription', function (error, lesson) {
      if ((error||{}).status === 401)
        db.put('ERROR', 'api getlessondescription 401 access denied', ER)
      else if ((error||{}).status === 404) {
        var max = STATUS.currentlesson-1
        if (STATUS.currentlesson < 0 ) STATUS.currentlesson = 0
        if (max <= 0 || STATUS.currentlesson <= 0) max = Infinity
        db.put('esova-maxlessons', max, ER)
      }
      else if (lesson) db.put('esova-lessonDescription', JSON.stringify(lesson), function (error) {
        if (error) db.put('ERROR', new Error('couldnt set lesson description'))
      })
      else {
        var creds = JSON.stringify({ status: 'logout' })
        function noop () { }
        db.put('esova-credentials', creds, noop)
        db.put('esova-currentLesson', '0', noop)
        db.put('esova-lesson', '0', noop)
      }
    })
  }
  function ER (error) {
    // @TODO
    // default error logging failed
    // => try to learn about error via ajax or some other technique
    // => if it fails inform user instead and hope he reports it in chat
    if (error) {
      console.error('couldnt store ERROR event')
      console.error(error)
    }
  }
  function onmouseover (event) {
    var key = this
    // STATUS = {}
    STATUS[key] = key+'--hover' // @HACK
    // db.put(key, key+'--hover', function (error) {
    //   if (error) db.put('ERROR', 'hover state could not be persisted', ER)
    // })
    renderSkillbox()
  }
  function onmouseout (event) {
    var key = this
    // STATUS = {}
    STATUS[key] = undefined // @HACK
    // db.del(key, function (error) {
    //   if (error) db.put('ERROR', 'unhover state could not be persisted', ER)
    // })
    renderSkillbox()
  }
}

///////////////////////////////////////////////
///////////////////////////////////////////////
///////////////////////////////////////////////
// 'use strict'
/******************************************************************************
  STACK BASE
******************************************************************************/
// var { h, hsvg } = require('_h')
// var s           = require('_s')
/******************************************************************************
  MAIN
******************************************************************************/
// module.exports = lessonHeaderContent

function lessonHeaderContent (db, engine) {


  /*****************************/
  // THEME + LOCAL CACHE + DEFAULT VALUES
  var esovagreen = '#35E38A'
  var esovablue = '#526E96'
  var STATUS = {
    // title: '...',
    // duration: '...'
  }
  // @TODO: switchboarded db should recognize all automatically
  var LOCALE = { // local CACHE - initial defaults are defined in template
    // "text/time": "min"
    "text/time": ""
  }

  /*****************************/
  // INITIALIZE STYLING
  var x = s(`
    lesson-name {
      color: ${esovablue};
    }
    lesson-duration {
      font-size: 20px;
      color: ${esovagreen};
    }
  `)


  /*****************************/
  // CHILD COMPONENTS & USED MODULES
  // ...


  /*****************************/
  // RENDER COMPONENT
  function renderLessonHeader () {
    var vtree = h('div', [
      h('div', { className: x('lesson-name') }, STATUS.title),
      h('div', { className: x('lesson-duration') }, [
        STATUS.duration + ' ' + LOCALE["text/time"]
      ])
    ])
    engine.push({
      type: 'vtree',
      key: 'lessonHeaderContent',
      value: vtree
    })
  }


  /*****************************/
  // INCOMING COMMUNICATION CHANNELS
  db.on('put', function (key, value) {
    try { value = JSON.parse(value) } catch (e) { value = defined(value, {}) }

    if (key === 'esova-lessonDescription') {
      STATUS.title = value.title
      STATUS.duration = value.duration
      renderLessonHeader()
    }
  })


  /*****************************/
  // INITIALIZE
  console.log('--------------------------------')
  console.log('INITIALIZE: lessonHeaderContent() component')
  console.log('--------------------------------')
  // @TODO: switchboarded db should recognize all automatically
  var FIELDS = Object.keys(LOCALE).length
  var normal = function () {
    engine.push({
      type: 'vtree',
      key: 'lessonHeaderContent',
      value: renderLessonHeader()
    })
  }
  var init = after(FIELDS, function (err) {
    if (err) db.put('ERROR', err)
    else {
      normal()
      init = normal
    }
  })
  db.on('put', function (key, value) {
    try { value = JSON.parse(value) } catch (e) { value = defined(value, {}) }

    if (LOCALE[key] !== undefined) {
      LOCALE[key] = value
      init()
    }
  })
  return undefined


  /*****************************/
  // HELPER
  // ...


}
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
// 'use strict'
/******************************************************************************
  STACK BASE
******************************************************************************/
// var { h, hsvg } = require('_h')
// var s           = require('_s')
/******************************************************************************
  CUSTOM
******************************************************************************/
var markdownStyles = document.createElement('style')
markdownStyles.innerHTML = `
  .markdown {
    padding: 5px;
  }
`
document.head.appendChild(markdownStyles)
/******************************************************************************
  MAIN
******************************************************************************/
// module.exports = lessonDescription

function lessonDescription (db, engine) {


  /*****************************/
  // THEME + LOCAL CACHE + DEFAULT VALUES
  var esovablue = '#526E96'
  var STATUS = {
    // lesson: '... loading lesson ...'
  }


  /*****************************/
  // INITIALIZE STYLING
  var x = s(`
    lesson-description {
      color: ${esovablue};
      font-size: 1em;
    }
  `)


  /*****************************/
  // CHILD COMPONENTS & USED MODULES
  // ...


  /*****************************/
  // RENDER COMPONENT
  function renderLessonDescription () {
    var vtree = h('div', {
      className: x('lesson-description') + '  markdown'
    }, STATUS.lesson)
    engine.push({
      type: 'vtree',
      key: 'lessonDescription',
      value: vtree
    })
  }


  /*****************************/
  // INCOMING COMMUNICATION CHANNELS
  db.on('put', function (key, value) {
    try { value = JSON.parse(value) } catch (e) { value = defined(value, {}) }

    if (key === 'esova-lessonDescription') {
      STATUS.lesson = JSON.parse(value.content)
      renderLessonDescription()
    }
  })


  /*****************************/
  // INITIALIZE
  console.log('--------------------------------')
  console.log('INITIALIZE: lessonDescription() component')
  console.log('--------------------------------')
  renderLessonDescription()
  return undefined


  /*****************************/
  // HELPER
  // ...
}
