// _vcs-esova/esova.js
'use strict'
/******************************************************************************
  STACK BASE
******************************************************************************/
const { h, hsvg } = require('_h')
const s           = require('_s')
/******************************************************************************
  STACK EXTEND
******************************************************************************/
const defined = require('defined')
/******************************************************************************
  CUSTOM
******************************************************************************/
const auth        = require('_auth')
const notFound    = require('_vcs-404')
/******************************************************************************
  MAIN
******************************************************************************/
module.exports = RootComponent

function RootComponent (db, engine) {


  /*****************************/
  // THEME + LOCAL CACHE + DEFAULT VALUES
  var REGISTERED    = false


  /*****************************/
  // INITIALIZE STYLING
  // ...


  /*****************************/
  // CHILD COMPONENTS & USED MODULES
  // @IDEA:
  // - child components have db listeners to act on
  // - parents inform world with db.put
  // IF: a parent requires a child, it for sure would like that child
  //     to act on certain actions and would like to learn about child actions
  // BECAUSE: without the direct "require" db.put/db.on communication would work
  // BUT: the "require" happens, because the parent outsources functionality
  // SO: how to auto-wire PARENT/CHILD after reqire?
  // ANSWER1: the child returns something (function OR object) as API
  // => parent needs to use API to tell child
  // => child uses API call from parent to return something
  // ANSWER2: parent alrady passes a customized db object to child
  // => whenever child "subscribes" or "puts", the parent can notice
  // => so parent can send to child & listen to child
  //   => thats important, because parent oursources
  // ....
  var renderMainPage = mainpage(db, engine)
  var renderNotFound = notFound(db, engine)


  /*****************************/
  // RENDER COMPONENT
  function render (error, update) {
    if (error) {
      var e = { name: error.name, message: error.message, stack: error.stack }
      db.put('ERROR', JSON.stringify(e, null, 2), function noop (err) { })
    }
    else renderMainPage()
  }


  /*****************************/
  // INCOMING COMMUNICATION CHANNELS
  if(REGISTERED) throw new Error ('this should never happen!')
  if (!REGISTERED) { // @HACK - why? what to do? wtf? :-)
    REGISTERED = true
    console.error('=== Global db listener activated! === ')
    db.on('put', function (key, value) {
      try { value = JSON.parse(value) } catch (e) { value = defined(value, {}) }

      if (key === 'target') {
        console.error('!!!! target !!!!!')
        // USE CASES: header.js on LOGOUT
        debugger
        // render()// @TODO: is a re-render here necessary?
      } else if (key === 'ERROR') {
        console.log('=====ERROR ORIGIN=====')
        console.log(value)
        console.log('=====ERROR ORIGIN=====')
      }
    })
  }


  /*****************************/
  // INITIALIZE
  console.log('--------------------------------')
  console.log('INITIALIZE: esova:RootComponent() component')
  console.log('--------------------------------')
  // @NOTE render() is triggered by router and pushes changes to engine$
  //       - can be triggered by others too
  // @TODO: render might be part of returned API of components if needed
  // @TODO: if not: remove render from component API
  // @TODO currently only used by raynos routes module
  // mainpage pushes to main engine to replace the whole site with error page
  // is currently only triggered by router and every time router navigates
  // to the routes that references "mainpage"
  return render
  // render() // because currently done by raynos router


  /*****************************/
  // HELPER
  // ...


}
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
// 'use strict'
/******************************************************************************
  STACK BASE
******************************************************************************/
// var { h, hsvg } = require('_h')
// var s           = require('_s')
/******************************************************************************
  MAIN
******************************************************************************/
// module.exports = footer

function footer (db, engine) {


  /*****************************/
  // THEME + LOCAL CACHE + DEFAULT VALUES
  // ...


  /*****************************/
  // INITIALIZE STYLING
  // ...


  /*****************************/
  // CHILD COMPONENTS & USED MODULES
  // ...


  /*****************************/
  // RENDER COMPONENT
  // ...


  /*****************************/
  // INCOMING COMMUNICATION CHANNELS
  // ...


  /*****************************/
  // INITIALIZE
  console.log('--------------------------------')
  console.log('INITIALIZE: footer() component')
  console.log('--------------------------------')
  engine.push({
    type: 'vtree',
    key: 'footer',
    value: ''
  })
  return undefined

  /*****************************/
  // HELPER
  // ...


}
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
// 'use strict'
/******************************************************************************
  STACK BASE
******************************************************************************/
// var { h, hsvg } = require('_h')
// var s           = require('_s')
/******************************************************************************
  CUSTOM
******************************************************************************/
const header      = require('_vcs-header')
const skillbox    = require('_vcs-skilltree/skillbox')
const promocode   = require('_vcs-promocode')
const lesson      = require('_vcs-lesson')
/******************************************************************************
  MAIN
******************************************************************************/
// module.exports = mainpage

function mainpage (db, engine) {


  /*****************************/
  // THEME + LOCAL CACHE + DEFAULT VALUES
  var VTREE = {
    header: '',
    content: '',
    footer: '',
    overlay: ''
  }
  var esovablue = '#526E96'
  var esovadarkblue = '#16243a'
  var STATUS = {
    // UI STATUS
    inactive: undefined,
    mode: 'skillbox' // maybe part of routing
  }


  /*****************************/
  // INITIALIZE STYLING
  var x = s(`
    app {
      position: relative;
      min-height: 100vh;
    }
    esova--inactive {
      filter: blur(2px);
      -webkit-filter: blur(2px);
    }
    esova {
      position: relative;
      display: flex;
      flex-direction: column;
      min-height: 100vh;
      width: 100%;
      background-color: ${esovablue};
    }
    header {
      min-height: 165px;
    }
    header--movie {
      background-color: ${esovadarkblue};
      box-shadow: 0px -3px 30px 0px hsla(0, 0%, 0%, 0.75);
      z-index: 1;
    }
    content {
      display: flex;
      flex-direction: row;
      justify-content: center;
      align-items: stretch;
      flex-grow: 1;
    }
    footer {
      min-height: 165px;
    }
    footer--movie {
      background-color: ${esovadarkblue};
    }
    overlay {
      position: absolute;
      top: 0;
      left: 0;
      display: flex;
      flex-direction: column;
      width: 100%;
      height: 100%;
    }
  `)


  /*****************************/
  // CHILD COMPONENTS & USED MODULES
  // @TODO
  // let them return stream$s that can be passed into hyperscript
  // so the custom VTREE CACHING becomes obsolete :-)
  promocode(db, { push: renderMainPage }) // show popup
  header(db, { push: renderMainPage })
  footer(db, { push: renderMainPage })

  skillbox(db, { push: renderMainPage })
  lesson(db, { push: renderMainPage })


  /*****************************/
  // RENDER COMPONENT
  function renderMainPage (/*error,*/ update) {
    // IDEA: old components should be able to answer with "no change"
    //       so that the last cached version can be used instead
    var childTree = (update||{}).value
    var childName = (update||{}).key
    // MAYBE UPDATE CHILDREN
    if (childName === 'skillbox') {
      VTREE.skillbox = childTree
    } else if (childName === 'lesson') {
      VTREE.lesson = childTree
    } else if (childName === 'footer') {
      VTREE.footer = childTree
    } else if (childName === 'header') {
      VTREE.header = childTree
      // OVERLAY
    } else if (childName === 'promocode' || childName === 'videopopup') {
      VTREE.overlay = childTree
    }
    // SELECT CHILDREN
    VTREE.content = STATUS.mode === 'lesson' ? VTREE.lesson : VTREE.skillbox

    // RENDER
    var vtree = h('div', { className: x('app') }, [
      h('div', { className: x('overlay') }, VTREE.overlay),
      h('div', {
        className: x('esova', STATUS.inactive ? 'esova--inactive' : undefined)
      }, [
        h('div', { className: x('header', STATUS.header) }, [
          VTREE.header
        ]),
        h('div', { className: x('content') }, VTREE.content),
        h('div', { className: x('footer', STATUS.footer) }, [
          VTREE.footer
        ])
      ])
    ])
    // PUSH UPDATE
    engine.push({
      type: 'vtree',
      key: 'root',
      value: vtree
    })
  }


  /*****************************/
  // INCOMING COMMUNICATION CHANNELS
  db.on('put', function (key, value) {
    try { value = JSON.parse(value) } catch (e) { value = defined(value, {}) }

    if (key === 'MODE') { // NAVIGATION: skillbox || lesson
      if (value === 'lesson') { // goto lesson
        // @TODO: this component is using "lesson" and "promocode"
        //        and "maincomponent", so it needs to orchestrate
        // @XXX: Alternative: lesson description can switch itself into lesson
        //      thus => lesson is listening to MODE
        STATUS.mode = 'lesson'
        STATUS.header = 'header--movie'
        STATUS.footer = 'footer--movie'
        renderMainPage()
      }
      else if (value === 'skillbox') {
        STATUS.mode = 'skillbox'
        STATUS.header = undefined
        STATUS.footer = undefined
        renderMainPage()
      }
      else { // @TODO: goto skillbox (skilltree)
        db.put('ERROR', 'MODE: '+value, ER)
      }
    }
    else if (key === 'esova-credentials' || key === 'OVERLAY') {
      var status = value.status
      var setERR = {
        'no valid token': true,
        'could not store credentials': true,
        'couldnt get "auth code" from github': true
      }
      var setINACTIVE = {
        'ok': true,
        'unmet requirements': true,
        'speed limit': true,
        'no valid promocode': true,
        'not ready': true
      }
      var setCLEAR = {
        'clear': true,
        'logout': true
      }

      if (setINACTIVE[status]) {
        STATUS.inactive = true
        renderMainPage()
      } else if (setCLEAR[status]) {
        STATUS.inactive = false
        renderMainPage()
      } else { // setERR or other
        db.put('ERROR', status, ER)
      }

    }
  })


  /*****************************/
  // INITIALIZE
  console.log('--------------------------------')
  console.log('INITIALIZE: mainpage() component')
  console.log('--------------------------------')
  db.get('esova-credentials', function (error, value) {
    try { value = JSON.parse(value) } catch (e) { value = defined(value, {}) }
    if (value.id) auth(db, function noop (error, help) { })
  })
  return renderMainPage


  /*****************************/
  // HELPER
  function ER (error) {
    // @TODO
    // default error logging failed
    // => try to learn about error via ajax or some other technique
    // => if it fails inform user instead and hope he reports it in chat
    if (error) {
      console.error('couldnt store ERROR event')
      console.error(error)
    }
  }


}
